cpp/util/ManagedArray.h

```diff
@@ -0,0 +1,146 @@
+ *  the original array alive if the original ManagedArray instances become
+ *  decoupled from it.
+ */
+template<typename T> class ManagedArray
+{
+public:
+    //! Default constructor with optional size for thread local arrays
+    /*! \param size Size of the array to allocate.
+     */
+    ManagedArray(unsigned int size=0)
+    {
+        m_size = std::make_shared<unsigned int>(size);
+        // Hard code making an array of at least size 1.
+        m_data = std::shared_ptr<std::shared_ptr<T> >(
+            new std::shared_ptr<T>(new T[1], std::default_delete<T[]>()));
+        create_new_array(size);
+        reset();
+    }
+
+    //! Destructor (currently empty because data is managed by shared pointer).
+    ~ManagedArray() {}
+
+    //! Update size of the array.
+    /*! \param size New size of the array.
+     */
+    void resize(unsigned int size)
+    {
+        if (size != *m_size)
+        {
+            *m_size = size;
+            reallocate();
+        }
+    }
+
+    //! Allocate new memory for the array.
+    /*! This function primarily serves as a way to reallocate new memory from
+     *  the Python API if needed.
+     * \param size New size of the array.
+     */
+    void reallocate()
+    {
+        create_new_array(*m_size);
+        reset();
+    }
+
+    //! Reset the contents of array to be 0.
+    void reset()
+    {
+        if (*m_size != 0)
+        {
+            memset((void*) get(), 0, sizeof(T) * (*m_size));
+        }
+    }
+
+    //! Return the underlying pointer (requires two levels of indirection).
+    T *get() const
+    {
+        std::shared_ptr<T> * tmp = m_data.get();
+        return (*tmp).get();
+    }
+
+    //! Writeable index into array.
+    T &operator[](unsigned int index)
+    {
+        if (index >= *m_size)
+        {
+            throw std::runtime_error("Attempted to access data out of bounds.");
+        }
+        return get()[index];
+    }
+
+    //! Read-only index into array.
+    const T &operator[](unsigned int index) const
+    {
+        if (index >= *m_size)
+        {
+            throw std::runtime_error("Attempted to access data out of bounds.");
+        }
+        return get()[index];
+    }
+
+    //! Get the size of the current array.
+    unsigned int size() const
+    {
+        return *m_size;
+    }
+
+    //! Make a deep copy of this array (i.e. copy the underlying data).
+    /*! This function returns by value since all memory ownership information
+     *  is transmitted via shared pointers. 
+     */
+    void *deepCopy()
+    {
+        ManagedArray<T> *deep_copy = new ManagedArray<T>(size());
+        // This *m_data will be a different shared_ptr than the current
+        // instance's, so assigning to it directly shouldn't affect any other
+        // arrays pointing to the memory of the current one and reallocating
+        // the memory of others won't affect this one.
+        *((*deep_copy).m_data) = *m_data;
```

Comment:
This is doing `shared_ptr = shared_ptr`, which makes a shallow copy of the data after all, right? Am I missing something here?

Reply:
Yeah, the function is named poorly, it's a shallow copy.

Just in case the purpose isn't obvious to everyone, let me explain so you all can provide some guidance on how I could improve the comments/docstrings or if we should maybe rework the current structure to be a bit more transparent as to why the code is designed the way it is (apologies in advance for the verbosity):

Normally a `ManagedArray`'s data is structured as `outer_ptr1->inner_ptr1->data1`. If you pass around `ManagedArray`s by value, you create other instances, where each instance has the structure `outer_ptrX->inner_ptr1->data`, which means that any of them can resize/reallocate/whatever the data and all the others will still be pointing to the right data (and same goes for updating the size).

Now, what we want is that if compute is called and there are numpy arrays referencing the current data, we need to keep that data exactly where it is because the numpy arrays are only passed the raw pointer to that data. The first line of `deepCopy` creates a new `ManagedArray` which now contains `outer_ptr2->inner_ptr2->data2`. The assignment above then sets `inner_ptr2 = inner_ptr1`, so the pointer sequence is now `outer_ptr2->inner_ptr2->data1` (and `data2` goes out of scope and is freed). We can now safely call `reallocate` on the original array, at which point we now have `outer_ptr[1|X]->inner_ptr1->data3` while the original `ManagedArrayManager` (which had its reference count incremented for every numpy array created) now contains the `ManagedArray` with the pointer sequence `outer_ptr2->inner_ptr2->data1`, so it's keeping the original data alive and is the only pointer to it.

Hopefully you already got that from the code and were just clarifying the poor naming, but if not there's my super long description.